docs/dev/relay_pass_manager.rst

```diff
@@ -0,0 +1,606 @@
+including the unused functions in the module. Note that this field is designed
+as a packed function, which enables the implementation of the optimization in
+both C++ and Python.
+
+Function-level Passes
+^^^^^^^^^^^^^^^^^^^^^
+
+Function-level passes are used to implement various intra function level
+optimizations for a given Relay module. It fetches one function at a time from
+the function list of a module for optimization and yields a rewritten Relay
+function. Most of Relay passes can be classified into this category, such as
+common subexpression elimination and inference simplification, etc.
+
+Note that the scope of passes at this level is a Relay function. Therefore, we
+cannot add or delete a function through these passes as they are not aware of
+the global information.
+
+.. code:: c++
+   
+    class FunctionPassNode : PassNode {
+      PassInfo pass_info;
+      runtime::TypedPackedFunc<Function(Function, Module, PassContext)> pass_func;
+      Module operator()(const Module& mod, const PassContext& pass_ctx) const final;
+      bool SkipFunction(const Function& func) const;
+      // Other members/methods are omitted...
+    };
+
+``pass_info`` is identical to what we just described in the module pass.
+``pass_func`` takes a function for optimization, it also needs a module as we
+may use it for reporting errors. A function could be annotated with
+"SkipOptimization" so that it will be ignored during optimization.
+
+Sequential Passes
+^^^^^^^^^^^^^^^^^
+
+SequentialPass is similar to Pytorch nn.Sequential that contains a host of
+passes for execution. It also has a disabled list that contains the list of
+disabled passes. For example, we may have a configuration which performs all
+passes at ``opt_level=3`` with some disabled passes using disabled_pass=xx
+provided by ``PassContext``. Now using ``SequentailPass``, we could glob all
+passes at ``opt_level=3`` and exclude those in disabled.
+
+.. code:: c++
+
+    class SequentialPassNode : PassNode {
+      PassInfo pass_info;
+      // Passes need to be executed.
+      Array<Pass> passes;
+      // Passes are disabled during optimization
+      std::vector<std::string> disabled;
+      bool PassEnabled(const PassInfo& info) const;
+      Module operator()(const Module& mod, const PassContext& pass_ctx) const final;
+    };
+
+Only a few passes currently in Relay is put in this group. For example,
+``FoldScaleAxis`` requires to dispatch ``ForwardFoldScaleAxis`` and
+``BackwardFoldScaleAxis`` internally. In addition, ``BackwardFoldScaleAxis`` is
+recommended to be fulfilled first. This pass, hence, is an ideal candidate for
+``SequentailPass``.
+
+The following code shows how individual passes in a sequential pass is invoked.
+Essentially, we sequentially execute each pass in a sequential pass using the
+order that they were appended to the pass list without the consideration of
+their orders. The pass phase ordering problem might be considered in the future.
```

Comment:
"using the order that they were appended to the pass list without the consideration of their orders" sounds like a contradiction. maybe trim off the end of the sentence?